### PR TITLE
Config loading

### PR DESCRIPTION
**Pull request by jonasfj**
Sorry, I had to touch `vendor/` again.. but it's not so bad as last time :)

Things worth looking at is `config/`.

### CODE REVIEW

File: config/provider.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// Providers returns a map of the registered TransformationProvider.
+func Providers() map[string]TransformationProvider {
+	mProviders.Lock()
+	defer mProviders.Unlock()
+
+	// Clone providers
+	m := map[string]TransformationProvider{}
+	for name, provider := range providers {
+		m[name] = provider
+	}
```
Comment (walac) [inline config/provider.go]: Why do we need to return a copy?

File: config/secrets/transform.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+func injectSecrets(
+	s *secrets.Secrets,
+	cache map[string]map[string]interface{},
+	val interface{},
+) (interface{}, error) {
+	switch val := val.(type) {
+	case []interface{}:
+		for i, v := range val {
+			v, err := injectSecrets(s, cache, v)
+			val[i] = v
```
Comment (walac) [inline config/secrets/transform.go]: Shouldn't we only assign the value if there is no error?

File: config/secrets/transform_test.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		},
+		Result: map[string]interface{}{
+			"key": "hello-world",
+			"credentials": map[string]interface{}{
+				"clientId":    "no-client",
+				"accessToken": "no-secret",
+			},
+			"secretsBaseUrl": s.URL,
+		},
+	}.Test()
```
Comment (walac) [inline config/secrets/transform_test.go]: We need a test case for array of interfaces, doesn't we?

File: config/configtest/transform.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+		panic(fmt.Sprintf("Unknown transform: %s", c.Transform))
+	}
+
+	err := transform.Transform(c.Input)
+	if err != nil {
+		panic(fmt.Sprintf("Transform(Input) failed, error: %s", err))
+	}
+
+	if !reflect.DeepEqual(c.Input, c.Result) {
+		panic(fmt.Sprintf("c.Result != c.Input, after transform, result: %#v ", c.Input))
```
Comment (jonasfj) [inline config/configtest/transform.go]: Probably, but I'm too lazy to carry around `testing.T`... It's just tests who cares about code quality.

File: config/env/transform_test.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	os.Setenv("CFG_TEST_ENV_VAR", "hello-world")
+	configtest.Case{
+		Transform: "env",
+		Input: map[string]interface{}{
+			"key": map[string]interface{}{"$env": "CFG_TEST_ENV_VAR"},
+		},
+		Result: map[string]interface{}{
+			"key": "hello-world",
+		},
+	}.Test()
```
Comment (jonasfj) [inline config/env/transform_test.go]: nope, good catch

File: config/provider.go (hunk 1 above)
Comment (jonasfj) [inline config/provider.go]: To avoid concurrency issues

File: config/secrets/transform.go (hunk 2 above)
Comment (jonasfj) [inline config/secrets/transform.go]: Sure, it's not important... I think modification of the input is undefined when there is an error.

File: config/secrets/transform_test.go (hunk 3 above)
Comment (jonasfj) [inline config/secrets/transform_test.go]: true

**Pull request merged**